### ISSUE TITLE
[RFC] Support MiniMax H3 T2VA FlowGRPO on GPU and NPU

### ISSUE THREAD

**Issue by mengchengTang** [verl-project/verl-omni#347]
# [RFC] Support MiniMax H3 T2VA FlowGRPO on GPU and NPU

## Feature request

Add LoRA FlowGRPO support for [`MiniMaxAI/MiniMax-H3`](https://huggingface.co/MiniMaxAI/MiniMax-H3) in `verl-omni`, covering both GPU and Ascend NPU.

The first version targets the FL2VA checkpoint in text-to-video-and-audio (`t2va`) mode:

- vLLM-Omni performs rollout;
- the FSDP Actor recomputes transition log-probabilities and updates Transformer LoRA weights;
- the text encoder, video VAE, and audio VAE remain frozen;
- PickScore evaluates prompt-video visual alignment over uniformly sampled frames;
- GPU and NPU use the same data, reward, and FlowGRPO contracts.

**Status:** proposal. vLLM-Omni supports H3 inference, but H3 FlowGRPO training has not been validated.

Related work:

- [MiniMax H3 official repository](https://github.com/MiniMax-AI/MiniMax-H3)
- [vLLM-Omni #5691: Add MiniMax H3 support](https://github.com/vllm-project/vllm-omni/pull/5691)
- [Diffusers #14355: Add MiniMax H3](https://github.com/huggingface/diffusers/pull/14355)

## Motivation

MiniMax H3 is a 33B dense Transformer that jointly generates video and stereo audio. Supporting it would extend `verl-omni` to another native audio-video policy and validate the framework on a substantially larger model.

H3 cannot directly reuse the existing single-latent FlowGRPO path:

1. Video and audio have separate latent states.
2. They use different sigma shifts: video `12.0`, audio `3.0`.
3. H3 uses data-pointing velocity, `x0 = xt + sigma * v`.
4. FlowGRPO requires stochastic trajectories and transition log-probabilities, while the current H3 inference scheduler does not return them.

The initial implementation should support GPU and Atlas A2/Ascend 910B NPU together, so backend differences are detected during development instead of added after the GPU path is complete.

### Goals

- T2VA LoRA FlowGRPO training.
- GPU rollout and FSDP training.
- NPU rollout and FSDP training.
- Frame-averaged PickScore reward.
- Rollout-replay log-probability monitoring.
- GPU/NPU precision comparison and short reward-curve validation.

### Non-goals

- First/last-frame conditioning and Ref2VA.
- H3-Context-IR and H3-Regenerate-2K.
- Full-parameter training.
- Additional sparse Attention, quantization, and performance tuning beyond the baseline FA3 path.
- Audio-quality, text-audio, and audio-video synchronization rewards.
- Training the text encoder or VAEs.

### Training flow

~~~text
prompt
  -> vLLM-Omni H3 rollout
  -> video/audio trajectories
  -> frame sampling + PickScore reward
  -> FlowGRPO advantage
  -> FSDP Actor replay
  -> LoRA update
  -> weight synchronization
~~~

### Model adapters

Add a model-specific pipeline:

~~~text
verl_omni/pipelines/minimax_h3_flow_grpo/
  common.py
  diffusers_training_adapter.py
  vllm_omni_rollout_adapter.py
~~~

The rollout adapter reuses vLLM-Omni's `MiniMaxH3Pipeline`. The training adapter uses the Diffusers H3 Transformer under FSDP. H3 model code should remain upstream rather than being copied into `verl-omni`.

### Trajectory and scheduler

The H3 adapter records:

- video/audio latent states and timesteps;
- per-modality transition log-probabilities;
- one normalized joint log-probability for FlowGRPO;
- prompt embeddings, masks, and packed-sequence metadata needed for replay.

The proposed joint log-probability is the latent-element-weighted average of the video and audio log-probabilities. Per-modality values remain available for debugging.

Add an H3-specific stochastic scheduler that:

- preserves H3's velocity definition;
- supports the two sigma schedules;
- matches the official deterministic update when stochastic noise is disabled;
- samples and replays transitions in FP32;
- returns video, audio, and joint log-probabilities.

### GPU and NPU

Both backends share the same adapter and scheduler logic. Backend-specific code is limited to device placement and Attention/operator selection.

| Backend | Actor Attention | Rollout Attention | Precision |
| --- | --- | --- | --- |
| GPU | `_flash_3_varlen_hub` | `FLASH_ATTN_3_HUB` | BF16 |
| NPU | `_native_npu` | `TORCH_SDPA` | BF16 |

GPU uses matched FA3 backends when supported. `native`/`TORCH_SDPA` remains the fallback for unsupported hardware or sequence-parallel configurations. The NPU path must preserve mask semantics, Ray device visibility, and rollout-replay consistency. Unsupported fused operators should fail clearly or use a documented native fallback.

### Reward
The first version uses PickScore as the only reward. It uniformly samples configurable frames from each generated video, scores each frame against the prompt, and averages the frame scores into one reward. Audio quality and audio-video synchronization are intentionally not optimized in the first version.

## Validation

1. **Contract tests**
   - trajectory shapes, dual sigma schedules, deterministic parity, and joint log-probability.
2. **GPU smoke**
   - real checkpoint rollout, one Actor update, LoRA synchronization, and finite PickScore rewards.
3. **NPU smoke**
   - the same rollout and Actor update on Ascend, with finite PickScore rewards and gradients.
4. **Train-rollout consistency**
   - report joint and per-modality `logprob_abs_diff` on GPU and NPU;
   - require `ratio_mean` near `1.0` and mismatch below the configured clipping range.
5. **GPU/NPU alignment**
   - use the same checkpoint, prompts, seeds, schedule, and LoRA configuration;
   - compare early-step losses, rewards, gradient norms, and log-probability differences;
   - confirm that short-run reward curves have the same overall trend.

## Dependencies and risks

| Item | Risk / handling |
| --- | --- |
| Diffusers H3 support | Pin a reviewed commit containing H3 and the required Attention backends |
| vLLM-Omni H3 NPU path | Pin a commit containing required H3 and NPU fixes |
| 33B Transformer memory | LoRA-only training, FSDP, offloading, and conservative rollout replicas |
| Dual scheduler correctness | Deterministic parity and replay tests are merge requirements |
| GPU FA3 compatibility | Validate `_flash_3_varlen_hub` against `FLASH_ATTN_3_HUB`; fall back to native/SDPA when unsupported |
| PickScore integration | The existing reward accepts one image; add video frame sampling and validate PickScore execution on the NPU path |
| PickScore coverage | Frame averaging does not evaluate motion, audio quality, or audio-video synchronization; document this first-version limitation |
| GPU/NPU numerical drift | Enable log-probability monitoring from the first runnable step |
| NPU fused Attention | Compare `_native_npu` with `native`; fall back to `native` if consistency does not pass |

## Implementation plan

- [ ] Add multimodal trajectory and H3 scheduler contracts.
- [ ] Add H3 Diffusers training adapter.
- [ ] Add vLLM-Omni rollout adapter and LoRA synchronization.
- [ ] Add video frame sampling, PickScore recipe, and prompt data example.
- [ ] Add GPU tests and a short training result.
- [ ] Add NPU tests, precision comparison, and a short training result.
- [ ] Document environment pins, resource requirements, and known limitations.

## Your contribution

I can implement the adapters, GPU/NPU recipes, tests, and validation experiments. Before opening a PR, I will repeat the repository's duplicate-work checks and include exact test commands/results, duplicate-work justification, and AI-assistance disclosure.

**Comment by SamitHuang** [comment]
Looks clear overall. Why not use FA3 for GPU actor/rollout attention backend?

**Comment by mengchengTang** [comment]
> Looks clear overall. Why not use FA3 for GPU actor/rollout attention backend?

Thanks for the suggestion. You are right that both the Actor and rollout paths support FA3, so matched FA3 backends are more appropriate for GPU. I will update the RFC to use _flash_3_varlen_hub for the Actor and FLASH_ATTN_3_HUB for rollout, with native/SDPA retained as a fallback.

**Comment by NancyFyong** [comment]
Hi @mengchengTang, I'm also interested in H3 integration; could we discuss and work on the integration together?

**Comment by mengchengTang** [comment]
> Hi [@mengchengTang](https://github.com/mengchengTang), I'm also interested in H3 integration; could we discuss and work on the integration together?

Hi @NancyFyong , just saw your message — thanks for reaching out! I'm currently working on the L2VA (FL2VA) pipeline and the PR at [#368](https://github.com/verl-project/verl-omni/pull/368) is still in draft. If you're interested in Ref2VA, feel free to build directly on top of my PR — that way you can reuse the shared H3 infrastructure (packed sequence layout, dual-sigma scheduling, agent loop, etc.) and avoid duplicating effort. If you'd like to sync more closely, we can exchange emails — happy to coordinate that way

**Comment by NancyFyong** [comment]
Thanks for the update. To avoid duplicating the T2VA work in #368, I want to record the scope and the material differences of the FL2VA work I am testing on my existing MiniMax-H3 branches.

- The NFT path extends #383's online direct-preference contract: rollout exports final clean target rows plus fixed first/last-frame condition state; the actor applies forward noise and loss only to generated video/audio target rows.
- The experimental FlowGRPO path uses separate video/audio SDE legs and configurable mean-reduced modality log-prob weights. FL2VA condition rows are pinned and excluded from both trajectories and log-probs. This differs from #368's current CPS/joint-weighted T2VA implementation.
- Both paths will consume vLLM-Omni's latest official FL2VA input matrix (`[0]`, `[-1]`, `[0, -1]`) rather than reimplementing request semantics in verl-omni.

I will not submit a duplicate of #368's T2VA base. The FlowGRPO work will remain scoped to the materially different replay/conditioning contract above, and I am happy to share the common H3 condition-layout tests/helpers so the implementations can converge where appropriate.
